### PR TITLE
[RFC] Clock stop and jack detection support 

### PR DESCRIPTION
**Pull request by RanderWang**
These patches  add support for clock stop mode and jack detection on soundwire platform.

### CODE REVIEW

File: drivers/soundwire/cadence_master.c [hunk 1]
```diff
@@ -1202,6 +1220,65 @@ static const struct sdw_master_port_ops cdns_port_ops = {
...
+	ret = sdw_bus_prep_clk_stop(&cdns->bus);
+	if (ret < 0)
+		return ret;
+
+	/* Enter clock stop */
+	ret = sdw_bus_clk_stop(&cdns->bus);
+	if (ret < 0)
+		return ret;
+
+	ret = cdns_set_wait(cdns, CDNS_MCP_STAT, CDNS_MCP_STAT_CLK_STOP);
```
Comment (plbossart) [inline drivers/soundwire/cadence_master.c:1258]: that part is suspicious.
When the bus changes to clock stop, it's an operation synchronous on all sides.
I don't see why we need to wait here, this should already be in CLK_STOP mode after calling sdw_bus_clk_stop().

Can you recheck @RanderWang?

File: drivers/soundwire/cadence_master.c [hunk 2]
```diff
@@ -1202,6 +1220,65 @@ static const struct sdw_master_port_ops cdns_port_ops = {
...
+	int ret;
+
+	/* Check suspend status */
+	status = cdns_readl(cdns, CDNS_MCP_STAT);
+	if (status & CDNS_MCP_STAT_CLK_STOP) {
+		dev_dbg(cdns->dev, "Clock is already stopped\n");
+		return 1;
+	}
+
+	/* Disable block wakeup */
```
Comment (plbossart) [inline drivers/soundwire/cadence_master.c:1240]: add a command that the hardware specification for the IP requires this bit to be set prior to entering Clock Stop state.

File: drivers/soundwire/intel.c [hunk 3]
```diff
@@ -1502,72 +1475,86 @@ static int intel_resume(struct device *dev)
...
 	if (ret) {
 		dev_err(dev, "%s failed: %d", __func__, ret);
 		return ret;
 	}
 
-	/* make sure all Slaves are tagged as UNATTACHED */
-	sdw_clear_slave_status(&sdw->cdns.bus);
+	if (clock_stop)
+		/* make sure all Slaves are tagged as UNATTACHED */
+		sdw_clear_slave_status(&sdw->cdns.bus);
```
Comment (plbossart) [inline drivers/soundwire/intel.c:1536]: No, this is not correct,
We should only tag the slaves as UNATTACHED when we do a hw_reset.

File: drivers/soundwire/intel.c [hunk 4]
```diff
@@ -1502,72 +1475,86 @@ static int intel_resume(struct device *dev)
...
-		dev_err(dev, "%s failed: %d", __func__, ret);
-		return ret;
-	}
+	link_flags = md_flags >> (sdw->cdns.bus.link_id * 8);
+	if (link_flags & SDW_INTEL_MASTER_DISABLE_CLOCK_STOP)
+		clock_stop = false;
 
-	/* make sure all Slaves are tagged as UNATTACHED */
-	sdw_clear_slave_status(&sdw->cdns.bus);
+	return _suspend(dev, clock_stop);
```
Comment (plbossart) [inline drivers/soundwire/intel.c:1579]: I suggest we first work through the code changes for suspend and pm_runtime, and at a later time we merge the code in helpers
This just too hard to review.

File: drivers/soundwire/bus.c [hunk 5]
```diff
@@ -686,6 +733,309 @@ static void sdw_modify_slave_status(struct sdw_slave *slave,
...
+
+	/* Now mark Slaves entering clock stop 1 as unattached */
+	list_for_each_entry(slave, &bus->slaves, node) {
+		if (!slave->dev_num)
+			continue;
+
+		mode = slave->curr_clk_stop_mode;
+
+		if (mode == SDW_CLK_STOP_MODE1)
+			sdw_modify_slave_status(slave, SDW_SLAVE_UNATTACHED);
```
Comment (plbossart) [inline drivers/soundwire/bus.c:954]: let's drop CLK_STOP_MODE1 for now, it's not really useful.

File: drivers/soundwire/bus.c [hunk 6]
```diff
@@ -753,6 +1103,46 @@ static int sdw_initialize_slave(struct sdw_slave *slave)
...
+			return ret;
+	}
+
+	/* Inform driver */
+	ret = sdw_slave_clk_stop_callback(slave, SDW_CLK_STOP_MODE1,
+					  SDW_CLK_POST_DEPREPARE);
+
+	return ret;
+}
+
```
Comment (plbossart) [inline drivers/soundwire/bus.c:1145]: let's drop clock_stop_mode1, it's not really useful

File: drivers/soundwire/cadence_master.c [hunk 7]
```diff
@@ -1278,6 +1282,53 @@ bool sdw_cdns_check_resume_status(struct sdw_cdns *cdns)
...
+ * sdw_cdns_resume: Cadence PM resume routine
+ *
+ * @cdns: Cadence instance
+ */
+int sdw_cdns_resume(struct sdw_cdns *cdns)
+{
+	int ret;
+
+	/* Disable block wakeup */
+	cdns_updatel(cdns, CDNS_MCP_CONTROL, CDNS_MCP_CONTROL_BLOCK_WAKEUP, 0);
```
Comment (plbossart) [inline drivers/soundwire/cadence_master.c:1295]: why is this here? It's already set on suspend?

File: drivers/soundwire/cadence_master.c [hunk 8]
```diff
@@ -1278,6 +1282,53 @@ bool sdw_cdns_check_resume_status(struct sdw_cdns *cdns)
...
+	/* commit changes */
+	cdns_update_config(cdns);
+
+	ret = cdns_clear_bit(cdns, CDNS_MCP_STAT, CDNS_MCP_STAT_CLK_STOP);
+	if (ret < 0) {
+		dev_err(cdns->dev, "Clear clock stop failed\n");
+		return ret;
+	}
+
+	ret = sdw_bus_exit_clk_stop(&cdns->bus);
```
Comment (plbossart) [inline drivers/soundwire/cadence_master.c:1325]: again that is suspicious. Why would we restart the clock before we tell the bus the clock restarted.

File: drivers/soundwire/intel.c [hunk 9]
```diff
@@ -408,6 +408,39 @@ static void intel_shim_wake(struct sdw_intel *sdw, bool wake_enable)
...
+			continue;
+
+		/*
+		 * wake up master and slave so that slave can notify master
+		 * the jack event and let codec driver check jack status
+		 */
+		list_for_each_entry(slave, &bus->slaves, node) {
+			if (slave->prop.wake_capable) {
+				pm_runtime_get(&slave->dev);
+				pm_runtime_put(&slave->dev);
```
Comment (plbossart) [inline drivers/soundwire/intel.c:437]: sorry, but this looks like a horrible hack.
what are you trying to achieve with this?

I suspect the idea is that we don't know which Slave device generated the wake, so we request all Slaves to become active and each driver will then double-check the interrupt source that caused the wake.

but isn't this what pm_runtime_request_resume() is all about

````
 int pm_request_resume(struct device *dev);
    - submit a request to execute the subsystem-level resume callback for the
      device (the request is represented by a work item in pm_wq); returns 0 on
      success, 1 if the device's runtime PM status was already 'active', or
      error code if the request hasn't been queued up
````

Comment (RanderWang) [reply drivers/soundwire/intel.c:437]: > sorry, but this looks like a horrible hack.
> what are you trying to achieve with this?
> 
> I suspect the idea is that we don't know which Slave device generated the wake, so we request all Slaves to become active and each driver will then double-check the interrupt source that caused the wake.

yes, this is my idea. we only know which master trigger a interrupt, but don't know which slave.
I will refine it in another way.

File: drivers/soundwire/bus.c (hunk 6 above)
Comment (RanderWang) [reply drivers/soundwire/bus.c:1145]: ok, I will refine it

File: drivers/soundwire/bus.c (hunk 5 above)
Comment (RanderWang) [reply drivers/soundwire/bus.c:954]: ok, I will refine it

File: drivers/soundwire/cadence_master.c (hunk 7 above)
Comment (RanderWang) [reply drivers/soundwire/cadence_master.c:1295]: yes, it is set in suspend.  Comments should be refined.

File: drivers/soundwire/cadence_master.c (hunk 8 above)
Comment (RanderWang) [reply drivers/soundwire/cadence_master.c:1325]: without master clock restart,  we can't tell the bus the clock restarted. I got IO transfer timeout.

File: drivers/soundwire/intel.c (hunk 3 above)
Comment (RanderWang) [reply drivers/soundwire/intel.c:1536]: yes, I got it.

File: drivers/soundwire/intel.c (hunk 4 above)
Comment (RanderWang) [reply drivers/soundwire/intel.c:1579]: yes, I will refine it

File: drivers/soundwire/cadence_master.c (hunk 2 above)
Comment (RanderWang) [reply drivers/soundwire/cadence_master.c:1240]: I will refine it

File: drivers/soundwire/cadence_master.c (hunk 1 above)
Comment (RanderWang) [reply drivers/soundwire/cadence_master.c:1258]: @plbossart   I checked internal code and spec again. Next step is to transfer ownership to glue hw, so we need to make sure master is in clk_stop mode now and do a smooth transition to glue hw.

Comment (plbossart) [reply drivers/soundwire/cadence_master.c:1258]: @RanderWang I don't mind if we check that the clock is stopped here, but the wait part should not happen. If we engage the clock stop procedure, and then keep the clock active for a bit longer until that bit is set, then I have no idea what slaves will do. It's actually an illegal transition, the Master *shall* drive the data line for 2 cycles before resuming the clock.
Let's recheck this, shall we?

**Comment by plbossart** [comment]
replaced by PR #1498
